File: JAVA/src/main/java/main/Person.java
package main;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Person implements PositionHolder {

  private int id;
  private double x;
  private double y;
  private double radius;
  private boolean infected;
  private double[] velocity;
  private int status;
  private boolean testing;
  private boolean testedPos;
  private boolean quarantined;
  private int collisionPartner;
  private double dateInfected;
  private double datePostIncub;
  private double dateResolution;

  public Person(int id) {
    this.id = id;
    this.x = Utils.randrange(Frame.getLeftWall() + Constants.PERSON_RAD,
        Frame.getRightWall() - Constants.PERSON_RAD);
    this.y = Utils.randrange(Frame.getTopWall() + Constants.PERSON_RAD,
        Frame.getBottomWall() - Constants.PERSON_RAD);
    this.radius = Constants.PERSON_RAD;
    boolean infected = Math.random() < Constants.START_INFECTED;
    if (infected) {
      this.startInfection();
    }
    this.updateInfection();
    velocity = new double[] {
        Utils.randrange(-Constants.PERSON_SPEED, Constants.PERSON_SPEED),
        Utils.randrange(-Constants.PERSON_SPEED, Constants.PERSON_SPEED)
    };
    setStatus(Constants.outcome.get("UNINFECTED"));
  }

  public int getId() {
    return id;
  }

  public int getStatus() {
    return status;
  }

  public void setStatus(int status) {
    this.status = status;
  }

  @Override
  public double getRadius() {
    return radius;
  }

  @Override
  public void setRadius(double radius) {
    this.radius = radius;
  }

  @Override
  public double getX() {
    return x;
  }

  @Override
  public void setX(double x) {
    this.x = x;
  }

  @Override
  public double getY() {
    return y;
  }

  @Override
  public void setY(double y) {
    this.y = y;
  }

  @Override
  public void incrementX(double x) {
    this.x += x;
  }

  @Override
  public void incrementY(double y) {
    this.y += y;
  }

  public boolean isTesting() {
    return testing;
  }

  public void setTesting(boolean testing) {
    this.testing = testing;
  }

  public boolean isInfected() {
    return infected;
  }

  public void setInfected(boolean infected) {
    this.infected = infected;
  }

  public boolean isTestedPos() {
    return testedPos;
  }

  public void setTestedPos(boolean testedPos) {
    this.testedPos = testedPos;
  }

  public boolean isQuarantined() {
    return quarantined;
  }

  public void takeTest() {
    if (this.infected && !this.testedPos) {
      this.testedPos = true;
      this.quarantined = true;
      Dashboard.incrementTotCases();
    }
  }

  public void move() {
    this.x += velocity[0];
    this.y += velocity[1];
    // horizontal boundary enforcement
    if (this.x + this.radius > Frame.getRightWall()) {
      this.x = Frame.getRightWall() - this.radius;
      this.velocity[0] *= Constants.FRICTION;
    } else if (this.x - this.radius < Frame.getLeftWall()) {
      this.x = Frame.getLeftWall() + this.radius;
      this.velocity[0] *= Constants.FRICTION;
    }

    // vertical boundary enforcement
    if (this.y + this.radius > Frame.getBottomWall()) {
      this.y = Frame.getBottomWall() - this.radius;
      this.velocity[1] *= Constants.FRICTION;
    } else if (this.y - this.radius < Frame.getTopWall()) {
      this.y = Frame.getTopWall() + this.radius;
      this.velocity[1] *= Constants.FRICTION;
    }

  }

  public void collide() {
    GameState gamer = Main.getGameState();
    int pop = gamer.getPopulation();
    List<Person> otherPeople = gamer.getPeople();
    for (int i = this.id + 1; i < pop; i++) {
      Person currentPerson = otherPeople.get(i);
      if (!(currentPerson.getStatus() == Constants.outcome.get("DEAD"))) {

        double dx = this.x - currentPerson.getX();
        double dy = this.y - currentPerson.getY();
        double distance = Math.sqrt(dx * dx + dy * dy);
        double minDist = this.radius + currentPerson.getRadius();

        if (distance < minDist) {
          double angle = Math.atan2(dy, dx);
          double targetX = this.x + Math.cos(angle) * minDist;
          double targetY = this.y + Math.sin(angle) * minDist;
          double ax = (targetX - currentPerson.getX()) * Constants.SPRING;
          double ay = (targetY - currentPerson.getY()) * Constants.SPRING;
          this.velocity[0] -= ax;
          this.velocity[1] -= ay;
          currentPerson.getVelocity()[0] += ax;
          currentPerson.getVelocity()[1] += ay;

          // End bounce
          if (!(this.quarantined || currentPerson.isQuarantined()
              || this.collisionPartner == currentPerson.getId())) {
            this.collisionPartner = currentPerson.getId();
            currentPerson.setCollsionPartner(this.id);
            Dashboard.increaseScore(10);
            if (Math.random() < 0.1) { // TODO make this a constant
              if (this.infected && !currentPerson.isInfected()) {
                currentPerson.startInfection();
              } else if (!this.infected && currentPerson.isInfected()) {
                this.startInfection();
              }
            }
          } else if (this.collisionPartner == currentPerson.getId()) {
            this.collisionPartner = -1; // unshaking of hands
            currentPerson.setCollsionPartner(-1);
          }
        }
      }
    }
  }

  private void startInfection() {
    Disease disease = Main.getDisease();
    this.infected = true;
    this.status = this.determineStatus(disease);
    this.dateInfected = Dashboard.getDaysElapsed();
    this.datePostIncub = this.dateInfected + disease.getIncubDays();
    this.dateResolution = this.dateInfected + disease.getDaysToResolution();

  }

  private int determineStatus(Disease disease) {
    double rand = Math.random();

    if (rand < disease.getAsymptProp()) {
      return 0;
    } else if (rand < disease.getAsymptProp() + disease.getMildProp()) {
      return 1;
    } else if (rand < disease.getAsymptProp() + disease.getMildProp()
        + disease.getSeriousNonDeathProp()) {
      return 2;
    } else {
      return 3;
    }
  }

  public void setCollsionPartner(int newCol) {
    this.collisionPartner = newCol;

  }

  public void updateInfection() {
    // TODO: something wrong
    if (Dashboard.getDaysElapsed() >= this.dateResolution) {
      // infection ended
      this.infected = false;
      this.dateInfected = -1;
      this.datePostIncub = -1;
      this.dateResolution = -1;

      // TODO: not working lol

      if (this.status == 3) {
        Dashboard.incrementTotDeaths(); // All deaths are tested.
        Dashboard.increaseScore(-100000);
      } else {
        if (this.testedPos) {
          Dashboard.incrementTotCases();
        }

        if (this.status == 2) {
          Dashboard.increaseScore(-10000);
        } else if (this.status == 1) {
          Dashboard.increaseScore(-1000);
        }
      }

      this.status = -1;
      this.testedPos = false;
      this.quarantined = false;
    } else {
      // infection ongoing
      if (this.testedPos) {
        Dashboard.incrementActiveCases();
      }

      Dashboard.incrementActualActive();

      if (this.status >= 2 && this.testedPos) {
        Dashboard.incrementSeriousCases();
      }
    }

  }

  @Override
  public double[] getVelocity() {
    return velocity;
  }

  @Override
  public void setVelocity(double[] newVel) {
    velocity = newVel;
  }

  public Map<String, Object> getPersonProperties() {
    Map<String, Object> mapper = new HashMap<>();
    mapper.put("id", this.id);
    mapper.put("x", this.x);
    mapper.put("y", this.y);
    mapper.put("r", this.radius);
    mapper.put("outcome", this.status);
    mapper.put("testedPos", this.testedPos);

    return mapper;
  }
}

File: static/js/person.js
// physics constants for move() and collide()
const FRICTION = -0.9;
const SPRING = 0.05;

//Instantiates the list of all people
Person.allPeople = [];

/**
 * @param {number} id 
 * @param {number} x the x-position relative to canvas (0, 0)
 * @param {number} y the y-position relative to canvas (0, 0)
 * @param {number} r the radius of the person measured in pixels
 * @param {boolean} infected the actual infection status, not the result of testing
 */

function Person(id, x, y, r, infected) {
  //Updates the list of all people
  Person.allPeople.push(this);

  // people
  this.id = id;

  // physics
  this.pos = createVector(x / scaler, y / scaler);
  this.r = r * scaler;
  this.vel = createVector(random(-1, 1), random(-1, 1));
  this.vel.setMag(PERSON_SPEED);

  /**
   * Use RNG to determine what the outcome of the infection will be.
   */
  this.infectionOutcome = function () {
    const rand = random();

    if (rand < disease.asymptProp) {
      return 0;
    } else if (rand < disease.asymptProp + disease.mildProp) {
      return 1;
    } else if (rand < disease.asymptProp + disease.mildProp + disease.seriousNonDeathProp) {
      return 2;
    } else {
      return 3;
    }
  }

  // infection and outcomes
  this.infected = infected;
  // outcomes: -1 = uninfected, 0 = asymptomatic, 1 = mild, 2 = serious, 3 = death
  this.outcome = this.infected ? this.infectionOutcome() : -1;
  this.dead = false;

  // infection timeline
  // Next line spreads out initial infection dates.
  this.dateInfected = this.infected ? random(-(disease.incubDays + disease.postIncubDays), 0) : null;
  this.datePostIncub = this.dateInfected + disease.incubDays; // date at the start of symptoms
  this.dateResolution = this.dateInfected + disease.daysToResolution;

  // interactions
  this.testing = false; // true if currently in contact with tester
  this.testedPos = false;
  this.quarantined = false;
  this.collidingWith = -1; // the index of current collision partner

  /**
   * This function displays the circles.
   * Colors are updated based on status.
   * In debug mode, calls displayLabel().
   */

  this.show = function () {
    fill(this.getColor());

    if (this.outcome >= 1){
      rect(this.pos.x, this.pos.y, this.r * 2, this.r * 2);
      //ellipse(this.pos.x, this.pos.y, this.r * 2, this.r * 2);
    } else{
      ellipse(this.pos.x, this.pos.y, this.r * 2, this.r * 2);
    }

    //@todo debug stuff.... change eventually
    if (DEBUG_MODE) {
      this.displayLabel();
    }
  }

  /**
   * This function displays hidden info on the person objects for debugging.
   */
  this.displayLabel = function () {
    // formatting
    textSize(SIMPLE_TEXT);
    textStyle(BOLD);
    fill(0);
    textAlign(CENTER, CENTER);

    let label = this.id;

    label = this.infected ? label + "\ninfected" + "\noutcome: " + this.outcome : label;
    label = this.testedPos ? label + "\ntestedPos" : label;
    label = this.quarantined ? label + "\nquarantined" : label;

    text(label, this.pos.x, this.pos.y);
  }

  /**
   * This is called when person contacts tester.
   * It updates status and dashboard variables if appropriate.
   */
  this.takeTest = function () {
    if (this.infected && !this.testedPos) {
      this.testedPos = true;
      this.quarantined = true;
      dash.totCases++;
    }
  }

  /**
   * This function updates position based on velocity.
   * It also enforces boundaries in the if () {} blocks.
   * FRICTION is a factor in the boundary enforcement.
   */
  this.move = function () {
    if (!this.dead && !this.quarantined) {
      this.pos.x += this.vel.x;
      this.pos.y += this.vel.y;
    }
    // horizontal boundary enforcement
    if (this.pos.x + this.r > rightWall) {
      this.pos.x = rightWall - this.r;
      this.vel.x *= FRICTION;
    } else if (this.pos.x - this.r < leftWall) {
      this.pos.x = leftWall + this.r;
      this.vel.x *= FRICTION;
    }

    // vertical boundary enforcement
    if (this.pos.y + this.r > bottomWall) {
      this.pos.y = bottomWall - this.r;
      this.vel.y *= FRICTION;
    } else if (this.pos.y - this.r < topWall) {
      this.pos.y = topWall + this.r;
      this.vel.y *= FRICTION;
    }
  }

  /**
   * This function controls collisions between people.
   * SPRING is a factor in change in velocity after collision.
   * Economic and disease activity happens at the end of the function.
   */
  this.collide = function () {
    // Look for collisions that are occurring.
    for (let i = this.id + 1; i < POPULATION; i++) {
      // Only collisions with living people occur.
      if (!Person.allPeople[i].dead) {
        let dx = Person.allPeople[i].pos.x - this.pos.x;
        let dy = Person.allPeople[i].pos.y - this.pos.y;
        let distance = sqrt(dx * dx + dy * dy);
        let minDist = Person.allPeople[i].r + this.r;

        // what to do if 2 people bump into each other
        if (distance < minDist) {
          let angle = atan2(dy, dx);
          let targetX = this.pos.x + cos(angle) * minDist;
          let targetY = this.pos.y + sin(angle) * minDist;
          let ax = (targetX - Person.allPeople[i].pos.x) * SPRING;
          let ay = (targetY - Person.allPeople[i].pos.y) * SPRING;
          this.vel.x -= ax;
          this.vel.y -= ay;
          Person.allPeople[i].vel.x += ax;
          Person.allPeople[i].vel.y += ay;

          // If both in collision are not quarantined, engage in trade and risk infection.
          // The last condition ensures these only execute once per collision, rather repeatedly during collision.
          if (!this.quarantined && !Person.allPeople[i].quarantined && this.collidingWith !== Person.allPeople[i].id) {
            // virtual shaking of hands
            this.collidingWith = Person.allPeople[i].id;
            Person.allPeople[i].collidingWith = this.id;

            dash.score += 10; // This reprsents a transaction.

            // Spreading of infection happens with probability of 0.1.
            if (random() < 0.1) { //@todo make this a constant
              if (this.infected && !Person.allPeople[i].infected) {
                Person.allPeople[i].startInfection();
              } else if (!this.infected && Person.allPeople[i].infected) {
                this.startInfection();
              }
            }
          }
        } else if (this.collidingWith === Person.allPeople[i].id) {
          this.collidingWith = -1; // unshaking of hands
          Person.allPeople[i].collidingWith = -1;
        }
      }
    }
  }
  
  this.startInfection = function () {
    this.infected = true;
    this.outcome = this.infectionOutcome();
    this.dateInfected = daysElapsed;
    this.datePostIncub = this.dateInfected + disease.incubDays;
    this.dateResolution = this.dateInfected + disease.daysToResolution
  }

  this.updateInfection = function () {
    if (daysElapsed >= this.dateResolution) {
      // infection ended
      this.infected = false;
      this.dateInfected = null;
      this.datePostIncub = null;
      this.dateResolution = null;

      if (this.outcome === 3) {
        this.dead = true;
        dash.totDeaths++; // All deaths are tested.
        dash.score -= 100000;
      } else {
        if (this.testedPos) {
          dash.totRecover++;
        }

        if (this.outcome === 2) {
          dash.score -= 10000;
        } else if (this.outcome === 1) {
          dash.score -= 1000;
        }
      }

      this.outcome = -1;
      this.testedPos = false;
      this.quarantined = false;
    } else {
      // infection ongoing
      if (this.testedPos) {
        dash.activeCases++;
      }

      dash.actualActive++;

      if (this.outcome >= 2 && this.testedPos) {
        dash.seriousCases++;
      }
    }
  }

  this.getColor = function(){
    if (this.dead) {
      return color(colors.DEAD); // @todo better visual
    } else if (this.testedPos) {
      return color(colors.QUARANTINED);
    } else {
      return color(colors.ALIVE);
    }
  }
}